### DISCUSSION TITLE
Connect from another python process rather than the VSCode debugger

### DISCUSSION THREAD

**Discussion by Ennosch** [microsoft/debugpy#1188] [General]
Hi,

I was wondering if one could connect to the debugger, using either the debupy cli or another python script, rather than VSCode.

Using:
- debugpy version: 1.6.5
- OS and version: macOS 11.6
- Python version Python 3.10.9
- VS Code or shell

I am running
listen.py
```python
import debugpy

debugpy.listen(("localhost", 5678))
debugpy.wait_for_client()
debugpy.breakpoint()
```
and try to connect via
```python
import debugpy

debugpy.connect(('localhost', 5678))
debugpy.wait_for_attach()
debugpy.breakpoint()
```
Will get me this error
```
Traceback (most recent call last):
  File "/Users/user/Dev/rez_dev/py310/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_comm.py", line 252, in _on_run
    self.process_net_command_json(self.py_db, json_contents)
  File "/Users/user/Dev/rez_dev/py310/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_process_net_command_json.py", line 172, in process_net_command_json
    assert request.type == 'request'
AssertionError
```

When trying to connect via
```
python -m debugpy --connect 5678 --pid 24017
```
I am getting another longer error.

Is this a trivial question ? Or would you expect this to be working without using any VSCode ?
It works fine when connecting via VSCode, as described in the docs.

**Comment by int19h** [comment]
debugpy is a debug adapter that implements the [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/). As such, it can be connected to by any DAP client or library that implements the protocol. Since debugpy itself is not a DAP client, it cannot connect to itself; it provides both `debugpy.connect()` and `debugpy.listen()` so that you can have better control over how the client is connected. You can tell debugpy to wait for an incoming connection with `listen()`, or you can tell it to open a connection to a client that is waiting for one with `connect()`.

I'm not familiar with third-party DAP client libraries, so I can't suggest any one in particular. The debugpy test suite has a very basic implementation of the client side, adequate for tests, which could be used as an example if you want to implement your own. Just bear in mind that anything in the `debugpy` package that is not exposed in `public_api.py` is considered an implementation detail that is subject to change without notice and should not be imported directly unless you're pinning an exact debugpy version.

**Reply by Ennosch** [reply]
Great. Thank you for clarifying @int19h. 

Continuing to use the VSCode extension for now.

> _The debugpy test suite has a very basic implementation of the client side, adequate for tests, which could be used as an example if you want to implement your own_

Will also take a look at that.

Meanwhile, please link to this discussion if anyones knows of such a DAP client that can be used instead of the VSCode extension.

**Reply by int19h** [reply]
https://github.com/puremourning/vimspector would be a third-party DAP client that's known to work with debugpy.
